### PR TITLE
Freddy/antehandler

### PR DESCRIPTION
**Pull request by freddyli7**
## Description

- Add module anteHandler
- Add checking logic for only existing module owner is able to add new module owner

## Changes

- 
- 
-

### CODE REVIEW

File: proto/chainlink/v1beta/genesis.proto [hunk 1]
```diff
@@ -1,12 +1,28 @@
...
+  bytes address = 1 [
+    (gogoproto.moretags) = "yaml:\"address\"",
+    (gogoproto.casttype) = "github.com/cosmos/cosmos-sdk/types.AccAddress"
+  ];
+  // pubKey defined the public key of the module owner
+  bytes pubKey = 2 [
+    (gogoproto.moretags) = "yaml:\"pub_key\""
+  ];
+  // the module owner who assigned this new module owner
+  bytes assignerAddress = 3 [
```
Comment (araskachoi) [inline proto/chainlink/v1beta/genesis.proto:24]: ok, i agree with this addition

File: x/chainlink/ante/ante.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+			authante.TxTimeoutHeightDecorator{},
+			authante.NewValidateMemoDecorator(ak),
+			authante.NewConsumeGasForTxSizeDecorator(ak),
+			authante.NewRejectFeeGranterDecorator(),
+			authante.NewSetPubKeyDecorator(ak), // SetPubKeyDecorator must be called before all signature verification decorators
+			authante.NewValidateSigCountDecorator(ak),
+			authante.NewDeductFeeDecorator(ak, bankKeeper),
+			authante.NewSigGasConsumeDecorator(ak, sigGasConsumer),
+			authante.NewSigVerificationDecorator(ak, signModeHandler),
+			authante.NewIncrementSequenceDecorator(ak),
```
Comment (araskachoi) [inline x/chainlink/ante/ante.go:37]: are all these required for the antehandler? or all needed to support

File: x/chainlink/ante/ante.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+	if err != nil {
+		return ctx, sdkerrors.Wrapf(sdkerrors.ErrLogic, "module owner check failed at anteHandler[ModuleOwnerDecorator]")
+	}
+
+	// no checking if module owner list is empty
+	if len(existingModuleOwnerList.GetModuleOwner()) == 0 {
+		return next(ctx, tx, simulate)
+	}
+
+	signers := make([]sdk.AccAddress, 0)
```
Comment (araskachoi) [inline x/chainlink/ante/ante.go:70]: are `signers` considered to be `moduleOwners`?

File: x/chainlink/ante/ante.go (hunk 2 above)
Comment (freddyli7) [reply x/chainlink/ante/ante.go:37]: those are sdk default anteHandler, all of them are required

File: x/chainlink/ante/ante.go (hunk 3 above)
Comment (freddyli7) [reply x/chainlink/ante/ante.go:70]: signer could be anyone

File: x/chainlink/types/msg.go [hunk 4]
```diff
@@ -86,7 +93,7 @@ func (m *ModuleOwner) GetSignBytes() []byte {
 }
 
 func (m *ModuleOwner) GetSigners() []sdk.AccAddress {
-	return []sdk.AccAddress{sdk.AccAddress(m.Address)}
+	return []sdk.AccAddress{sdk.AccAddress(m.AssignerAddress)}
```
Comment (araskachoi) [inline x/chainlink/types/msg.go:96]: does the `assigner` become the `signer` of the newly added moduleowner then?

**Review by araskachoi — APPROVED** [review]

File: x/chainlink/types/msg.go (hunk 4 above)
Comment (freddyli7) [reply x/chainlink/types/msg.go:96]: yes, the assigner is always the signer for module owner tx who ideally should be one of the existing module owner

**Pull request merged**